Preliminary attempt to modernize to pymc4/5 and more recent matplotlib

terrycojones opened the pull request:
Hi @treszkai

On the weekend I tried to get BEST working with a current pymc and matplotlib. I was only using `best.analyze_two` and `best.plot_all`. I really didn't know what I was doing almost all the time, so some (maybe many) of the changes in this pr might be quite wrong. But it's a start. The code does run on my machine and the plots produced look to be ok. It will be quite some work to get this pr to a state where it could be merged (supposing you want that at some point).

If you're still interested in having a pymc4+ version, I'm happy to do the work but will for sure need some guidance. One question is whether you would still want to support pymc3. If not, we can set a minimum version in `setup.py` and then simplify some code. Same goes for versions of arviz and matplotlib. My inclination would be to just bring them all up to date and tell people that if they have to use old versions of these libraries then they should install an earlier version of BEST. But that's slightly user-unfriendly.

Anyway, thanks for the nice code! I almost never come across Python that I consider as well written :-).  BTW, have you considered running `black` on the code?

I haven't done anything with the tests here, and they fail on my machine.  BTW, I am also in Berlin.

terrycojones committed:
Preliminary attempt to modernize to pymc4/5 and more recent matplotlib

treszkai commented:
Hello @terrycojones, thanks for the pull request!

I'll check it until Friday and answer your questions. It was a conscious decision two years ago not to use ArviZ, but I'll revisit the project now and see your changes. Same goes for pymc4, which was in very early beta at the time.

Yes, seeing that there are now contributions every now and then, an autoformatter would be reasonable.

terrycojones commented:
OK, sounds good. Re arviz, I really had very little idea what I was doing. I was running with 4 chains and got a message telling me that the data being passed to the function you were originally calling was the wrong shape. I looked around online and saw this `arviz.extract` function, which somehow combines data from all chains, and gave it a try. It "worked" (in the sense that the error message went away), and I kept going, figuring that if something was really wrong I would see it in the eventual figures.  I can't emphasize enough that I don't know what I'm doing! Trying to learn, though. Thanks :-)

treszkai commented on best/model.py:
```diff
@@ -62,26 +62,32 @@ def sample(self, n_samples: int, **kwargs) -> MultiTrace:
-        pm_major, pm_minor, *_ = pm.__version__.split('.')
-        if (int(pm_major), int(pm_minor)) < (3, 7):
+        pm_major, pm_minor = map(int, pm.__version__.split('.')[:2])
+        if (pm_major, pm_minor) < (3, 7):
```
I prefer the original. (Also later at the ArviZ version comparison.)

treszkai commented on best/model.py:
```diff
@@ -62,26 +62,32 @@ def sample(self, n_samples: int, **kwargs) -> MultiTrace:
-                    kwargs['tune'] = 2000
-                    print('\nDue to potentially incorrect estimates, rerunning sampling '
-                          'with {} tuning samples.\n'.format(kwargs['tune']), file=sys.stderr)
+        if pm_major < 4:
```
We can now drop this and require PyMC 5.x.

treszkai commented on best/model.py:
```diff
@@ -62,26 +62,32 @@ def sample(self, n_samples: int, **kwargs) -> MultiTrace:
+                              file=sys.stderr)
+                    else:
+                        print('\nThe samples maybe are still not totally okay. '
+                              'Try rerunning the analysis.')
```
Is it possible to check whether the simulation was successful in pymc 5? We'd really need something like that.

treszkai commented on best/model.py:
```diff
@@ -287,9 +293,13 @@ def hdi(self, var_name: str, credible_mass: float = 0.95):
-        if (int(az_major), int(az_minor)) >= (0, 8):
-            return tuple(arviz.hdi(self.trace[var_name], hdi_prob=credible_mass))
+        az_major, az_minor = map(int, arviz.__version__.split('.')[:2])
+        if (az_major, az_minor) >= (0, 8):
```
Again, we can just assume that Arviz has minor version at least 8, so we can delete the else branch here.

treszkai commented on best/model.py:
```diff
@@ -287,9 +293,13 @@ def hdi(self, var_name: str, credible_mass: float = 0.95):
+        az_major, az_minor = map(int, arviz.__version__.split('.')[:2])
+        if (az_major, az_minor) >= (0, 8):
+            hdi = arviz.hdi(self.trace, var_names=[var_name], hdi_prob=credible_mass)
+            if isinstance(hdi, np.narray):
```
I suppose you mean `np.ndarray` here.

But either way, this `hdi` variable should have a given type, so either the `if` or the `else` branch should be deleted.

treszkai commented on best/plot.py:
```diff
@@ -253,9 +254,9 @@ def plot_data_and_prediction(best_results: BestResults,
+        nus = arviz.extract(trace, var_names=['Normality'])
     elif isinstance(best_results, BestResultsOne):
         means = trace['Mean']
         sigmas = trace['Sigma']
```
Could you update the code for single-group analysis as well?

treszkai commented:
Hello again,

I broke my wrist, sorry about the delay. Thanks for the initiative, I'd like to get this project up and running again.

I like the changes and I just made some minor comments. Haven't tested your code myself yet, though. When you say it's good, then I will do so, check the changes again and merge this PR.

> One question is whether you would still want to support pymc3. If not, we can set a minimum version in setup.py and then simplify some code.

Yeah, just ditch support for the old versions, as a rule of thumb, anything older than a year should go. If someone's upgrading their version of `best`, they could upgrade their other packages as well. Say, pymc 5.x and  arviz >= 0.15.0.

I suppose you are aware that there are still references to buy PyMC3 throughout the code.

Oh hello Berliner! What a coincidence, super cool.

Best,
Laszlo

treszkai commented:
And yes to adding Black to the pre-commit config, but I would do that on a separate branch once this is merged.

terrycojones commented:
Hi. I only just saw this - thanks for the reply! I will put it on my list of stuff to do. I had a hip replacement, so don't worry about the slowness :-) I hope the wrist is going well. I broke one once too, some decades ago and it wasn't much fun.

terrycojones committed:
Made tests work. Removed some PyMC3 code relevant only to earlier versions. Small clean-ups.

terrycojones commented:
Hi again. I finally made some more time to push this forward. The tests work now. I changed the caching save/restore code in the tests a bit as the pickle was failing and I didn't immediately see how to fix it. I also changed the suffix on the saved files to `.cdf` (the format stored is `netcdf`) to avoid the issue of trying to load an old pickle file from a pre-existing `.bin` pickle file.  Comments welcome, of course. I still don't really know what I'm doing.

I removed some code from `best/__init__.py` that I think was only relevant some years ago to old numpy. And some other code testing for old versions of PyMC3.

The test that the `tests/data` dir exists & creating it if not isn't needed, because that dir is in the repo, right?

terrycojones commented:
Hi again. I've been away in Australia (family medical issues) but am now back in Berlin. And happy to help keep moving this along :-)

treszkai commented on best/model.py:
```diff
@@ -17,10 +17,15 @@
+from arviz import InferenceData
 import scipy.stats as st
 
+# Ensure PyMC3 is version >= 5.
+if int(pm.__version__.split('.')[0]) < 5:
+    raise ImportError('BEST requires a PyMC3 version >=5')
+
```
```suggestion
# Ensure PyMC is version >= 5.
if int(pm.__version__.split('.')[0]) < 5:
    raise ImportError('BEST requires PyMC version >=5')
```

treszkai commented on best/model.py:
```diff
@@ -287,11 +276,11 @@ def hdi(self, var_name: str, credible_mass: float = 0.95):
-            return tuple(arviz.hdi(self.trace[var_name], hdi_prob=credible_mass))
+        hdi = arviz.hdi(self.trace, var_names=[var_name], hdi_prob=credible_mass)
+        if isinstance(hdi, np.ndarray):
+            return tuple(hdi)
         else:
-            return tuple(arviz.hpd(self.trace[var_name], credible_interval=credible_mass))
+            return hdi[var_name].values[0], hdi[var_name].values[1]
```
`arviz.hdi` always returns an `xarray.Dataset` (because `self.trace` is of type `InferenceData`), so the first branch is not necessary, and lines 281..283 can be replaced with the following:

```python
    return tuple(hdi[var_name].values)
```

treszkai commented on best/model.py:
```diff
@@ -355,7 +343,7 @@ def posterior_mode(self,
             The posterior mode.
         """
-        samples = self.trace[var_name]
+        samples = arviz.extract(self.trace.posterior[var_name]).to_dataarray()
```
This could be written more concisely as

```suggestion
        samples = arviz.extract(self.trace.posterior, var_names=var_name)
```

treszkai commented on best/plot.py:
```diff
@@ -88,9 +89,9 @@ def plot_posterior(best_results: BestResults,
         >>> plt.show()
     """
-    samples = best_results.trace[var_name]
+    samples = best_results.trace.posterior[var_name]
```
This doesn't work because the samples from the 4 chains are in different dimensions so the histogram looks messed up, so we need to flatten them like before:

```suggestion
    samples = arviz.extract(best_results.trace.posterior, var_names=var_name)
```

treszkai commented on setup.py:
```diff
@@ -22,10 +22,10 @@
         'scipy',
         'matplotlib>=3.0.0',
-        'pymc3<4.0',
+        'pymc3>=5',
```
```suggestion
        'pymc>=5',
```

treszkai commented on tests/test_plots.py:
```diff
@@ -29,23 +28,22 @@ def test_plot_all_one(data_dir):
         print('Saving results to "%s"' % filename)
-        with open(filename, "wb") as f:
-            pickle.dump(best_out, f)
+        best_out._trace.to_netcdf(filename)
```
Let's use the read-only property here that's meant for public access:

```suggestion
        best_out.trace.to_netcdf(filename)
```

treszkai commented on tests/test_plots.py:
```diff
@@ -58,23 +56,22 @@ def test_plot_all_two(data_dir):
         print('Saving results to "%s"' % filename)
-        with open(filename, "wb") as f:
-            pickle.dump(best_out, f)
+        best_out._trace.to_netcdf(filename)
```
```suggestion
        best_out.trace.to_netcdf(filename)
```

treszkai commented on tests/test_plots.py:
```diff
@@ -88,20 +85,19 @@ def test_plot_posterior(data_dir):
         print('Saving results to "%s"' % filename)
-        with open(filename, "wb") as f:
-            pickle.dump(best_out, f)
+        best_out._trace.to_netcdf(filename)
```
```suggestion
        best_out.trace.to_netcdf(filename)
```

treszkai commented on best/model.py:
```diff
@@ -55,33 +60,17 @@ def observed_data(self, group_id: int):
-                          'Try rerunning the analysis.')
+        kwargs.setdefault('target_accept', 0.9)
+
+        with self.model:
+            trace = pm.sample(n_samples, **kwargs)
```
This change is my one big problem with this PR. It completely drops checking for `trace.report.ok`, which sometimes is necessary. (To simulate such a case, run `best_out = best.analyze_two(group1_data, group2_data, tune=0, n_samples=1000, keep_warning_stat=True)`.)

After hours of checking your changes and debugging `InferenceData`, this is the last line that I came back to. The obvious solution is to add `return_inferencedata=False` to `pm.sample`, which would keep everything as a `MultiTrace` instead of the new `InferenceData`, so we could check `trace.report.ok`. I see no problems with that, but we'd need to drop much of our adaptations from this PR.

Alternatively, we could add `keep_warning_stat=True` and replace `trace.report.ok` with `trace.sample_stats['diverging'].any()`. This would prevent the serialisation to NetCDF, which we rely on in the tests. I don't see too many reasons to use `InferenceData` instead of `MultiTrace` (I liked this pymc-internal format more), apart from keeping these changes that you (and I) worked on in this PR – which is not an amazing reason tbh.

Do you have stronger reasons for going with `InferenceData`?

treszkai replied:
Okay, let's go with `InferenceData`, I like that the user of BEST doesn't need to care about this PyMC-internal data structure (even though I do like it).

If you are up for making the changes, let me know until the end of the week. Thanks for your work so far.

terrycojones replied:
Hi. I only just saw this. I'm up for doing the work, but don't have time right now to think. I will look on the weekend. Thanks for putting in the hours to look at the pr!

treszkai replied:
Hello, if you can do these changes by the end of this week, then I could review and merge them next week, otherwise it would need to wait another week while I'm AFK.

Thanks!
Laszlo

treszkai replied:
Hey @terrycojones, any updates on this? I would fix it around Friday if you aren't (which is fine, you'd still get credited, I just don't want to drag this any further).

treszkai reviewed (changes requested):
Hi,

Sorry for the months of delay, I just got back into developer mode after a long hiatus. If you don't care anymore about the project, no hard feelings, let me know.

Thanks again for the proposed changes! They mostly work but there are some changes that I detailed in the comments. The single showstopper is in [best/model.py:L73](https://github.com/treszkai/best/pull/8/files/d2dbc4bb81414978bbc84db1b74b87f84800c295#r1603608945), so check that comment first.

Best,
Laszlo

treszkai commented:
Closed in favor of https://github.com/treszkai/best/pull/10. @terrycojones thanks for kick-starting this issue!
